File: assets/ACS2/Classifier.cc
/*
/       ACS2 in C++
/	------------------------------------
/       choice of without/with GA, subsumption, PEEs
/
/     (c) by Martin V. Butz
/     University of Wuerzburg / University of Illinois at Urbana/Champaign
/     butz@illigal.ge.uiuc.edu
/     Last modified: 02-23-2001
/
/     classifier class.
*/

#include<iostream>
#include<assert.h>

#include"Condition.h"
#include"Action.h"
#include"Effect.h"
#include"PMark.h"
#include"Classifier.h"

/**
 * Constructs new empty classifier with specified action.
 */
Classifier::Classifier(Action *action) {
    initialize(new Condition(), new Action(action), new Effect(), new PMark(), 0, Q_INI, R_INI, IR_INI, AVT_INI);
}

/**
 * Copies complete classifier with all components except for the mark!
 */
Classifier::Classifier(Classifier *cl, int time) {
    initialize(new Condition(cl->C), new Action(cl->A), new Effect(cl->E), new PMark(), time, cl->q, cl->r, cl->i,
               cl->tav);
}

/**
 * Creates a classifier that anticipates the change correctly.
 */
Classifier::Classifier(Perception *p0, Action *act, Perception *p1, int time) {
    E = new Effect();
    C = E->getAndSpecialize(p0, p1);
    initialize(C, new Action(act), E, new PMark(), time, Q_INI, R_INI, IR_INI, AVT_INI);
}

/**
 * Controls the expected case of a classifier.
 */
Classifier *Classifier::expectedCase(Perception *percept, int time) {
    E->updateEnhancedEffectProbs(percept, BETA);

    //Return if the specificity threshold is already reached
    int noSpec = getUnchangeSpec();

    //Check out the possible specializations comparing mark to the percept
    Condition *diff = M->getDifferences(percept);
    if (diff == 0) {
        //No possibility for a further specialization was found
        if (DO_PEES && !M->isEmpty()) {
            ee = 1;
        }
        increaseQuality();
        return 0;
    }

    //Create a new classifier
    Classifier *cl = new Classifier(this, time);

    int noSpecNew = diff->getList()->getSize();

    if (noSpec >= U_MAX) {
        //Remove positions if the classifier is too specific anyways
        while (noSpec >= U_MAX) {
            while (!cl->generalizeRandomUnchangeCond(noSpec));//Remove definitely one of the old specializations
            noSpec--;
        }
        while (noSpec + noSpecNew > U_MAX) {
            if (frand() < 0.5) {
                if (diff->generalize((int) (frand() * noSpecNew)))
                    noSpecNew--;
            } else {
                if (cl->generalizeRandomUnchangeCond(noSpec))
                    noSpec--;
            }
        }
    } else {
        //Remove positions if toxo many specializations are suggested
        while (noSpec + noSpecNew > U_MAX) {
            if (diff->generalize((int) (frand() * noSpecNew)))
                noSpecNew--;
            else
                cout << "Less specialization in expected case should always work!?" << endl;
        }
    }

    if (!cl->C->specialize(diff)) {
        cout << "Mistake in the expected case during specialization!" << endl;
        exit(0);
    }
    delete diff;
    if (cl->q < 0.5)
        cl->q = 0.5;

    return cl;
}

/**
 * Generalizes randomly one unchanging attribute in the condition. 
 * An unchanging attribute is one that is anticipated not to change in the effect part.
 */
int Classifier::generalizeRandomUnchangeCond(int noSpec) {
    int pos = (int) (frand() * noSpec); //determine which unchanging attribute to generalize

    CharPosList *cpl = C->getList();
    cpl->reset();
    ProbCharPosList *epl = E->getList();
    epl->reset();

    CharPosItem *cpi = cpl->getNextItem();
    ProbCharPosItem *epi = epl->getNextItem();
    while (cpi != 0 && epi != 0) {
        if (cpi->getPos() < epi->getPos()) {
            if (pos == 0)
                break;
            pos--;
            cpi = cpl->getNextItem();
        } else if (cpi->getPos() == epi->getPos()) {
            if (epi->getItem()->doesContain(cpi->getChar())) {//explicit equal in enhanced effect part
                if (pos == 0)
                    break;
                pos--;
            }
            cpi = cpl->getNextItem();
            epi = epl->getNextItem();
        } else {
            epi = epl->getNextItem();
        }
    }

    if (cpi != 0) {
        C->generalize(cpi);
        return 1;
    }
    return 0;
}

/**
 * Returns is the Effect part contains specified attributes.
 */
int Classifier::doesAnticipateChange() {
    return E->getSpecificity();
}

/**
 * Determines the number of specified unchanging attributes in the classifier.
 * An unchanging attribute is one that is anticipated not to change in the effect part.
 */
int Classifier::getUnchangeSpec() {
    CharPosList *cpl = C->getList();
    cpl->reset();
    ProbCharPosList *epl = E->getList();
    epl->reset();
    int spec = 0;

    CharPosItem *cpi = cpl->getNextItem();
    ProbCharPosItem *epi = epl->getNextItem();
    while (cpi != 0 && epi != 0) {
        if (cpi->getPos() < epi->getPos()) {
            spec++;
            cpi = cpl->getNextItem();
        } else if (cpi->getPos() == epi->getPos()) {
            if (epi->getItem()->doesContain(cpi->getChar()))//explicit equal in enhanced effect part
                spec++;
            cpi = cpl->getNextItem();
            epi = epl->getNextItem();
        } else {
            epi = epl->getNextItem();
        }
    }
    while (cpi != 0) {
        spec++;
        cpi = cpl->getNextItem();
    }

    return spec;
}

/**
 * Controls the unexpected case of a classifier.
 * Returns a specialized classifier if generation was possible, 0 otherwise.
 */
Classifier *Classifier::unexpectedCase(Perception *p0, Perception *p1, int time) {
    decreaseQuality();
    mark(p0);

    if (E->isSpecializable(p0, p1) != 0) {
        Classifier *cl = new Classifier(this, time);

        Condition *diff = cl->E->getAndSpecialize(p0, p1);
        cl->C->specialize(diff);

        delete diff;

        if (cl->q < 0.5)
            cl->q = 0.5;

        return cl;
    }

    return 0;
}

/**
 * Returns if the classifier is equal to classifier cl in condition, action and effect part.
 */
int Classifier::isSimilar(Classifier *cl) {
    if (C->isEqual(cl->C) && A->isEqual(cl->A) && E->isEqual(cl->E))
        return 1;
    return 0;
}

/**
 * Returns if the classifier subsumes classifier cl.
 */
int Classifier::doesSubsume(Classifier *cl) {
    if (isSubsumer() && isMoreGeneral(cl) && C->doesMatch(cl->C) && A->isEqual(cl->A) && E->isEqual(cl->E))
        return 1;
    return 0;
}

/**
 * Returns if the classifier matches the percept.
 */
int Classifier::doesMatch(Perception *percept) {
    return C->doesMatch(percept);
}

/**
 * Returns if the classifier correctly anticipates the change from p0 to p1.
 */
int Classifier::doesAnticipateCorrect(Perception *p0, Perception *p1) {
    return E->doesAnticipateCorrectly(p0, p1);
}

/**
 * Returns the anticipation, the classifier believes to happen most probably.
 * This is usually the normal anticipation. However, if PEEs are activated, the most probable
 * value of each attribute is returned.
 */
Perception *Classifier::getBestAnticipation(Perception *percept) {
    return E->getBestAnticipation(percept);
}

/**
 * Returns the backwards anticipation.
 * Returns 0 if the backwards anticipation was impossible to create! This is the case if
 * changing attributes are not specified in the conditions.
 */
Perception *Classifier::getBackwardsAnticipation(Perception *percept) {
    Perception *p = C->getBackwardsAnticipation(percept);
    if (!E->doesSpecifyOnlyChangesBackwards(p, percept)) {
        //If a specified attribute in the effect part matches the anticipated p, the backwards anticipation fails
        // (because a specified attribue in E means a change!).
        delete p;
        return 0;
    }
    return p;
}

/**
 * Returns if 'percept' is matched by the anticipations.
 * This is only the case if the specified conditions that have '#'-symbols in the effect part are also matched!
 */
int Classifier::doesMatchBackwards(Perception *percept) {
    Perception *p = C->getBackwardsAnticipation(percept);
    if (E->doesMatch(percept, p)) {
        delete p;
        return 1;
    }
    delete p;
    return 0;
}

int Classifier::doesLink(Classifier *cl) {
    CharPosList *cpl = C->getList();
    ProbCharPosList *pcpl = E->getList();
    CharPosList *cpl2 = cl->C->getList();
    ProbCharPosList *pcpl2 = cl->E->getList();

    cpl->reset();
    cpl2->reset();
    pcpl->reset();
    pcpl2->reset();

    //Check if cl.C matches this classifier
    CharPosItem *cpi = cpl->getNextItem();
    ProbCharPosItem *pcpi = pcpl->getNextItem();
    CharPosItem *cpi2 = cpl2->getNextItem();
    ProbCharPosItem *pcpi2 = pcpl2->getNextItem();

    while ((cpi2 != 0 || pcpi2 != 0) && (cpi != 0 || pcpi != 0)) {
        if (cpi != 0) {
            if (pcpi == 0 || cpi->getPos() < pcpi->getPos()) {
                if (!doesLink(cpi->getChar(), cpi->getPos(), cpl2, &cpi2, pcpl2, &pcpi2))
                    return 0;
                cpi = cpl->getNextItem();//this item is OK -> get next one
            } else {//given: pcpi->getPos() is the next existent, relevant item
                if (!doesTightLink(pcpi->getItem()->getBestChar(), pcpi->getPos(), cpl2, &cpi2))
                    return 0;
                if (cpi->getPos() == pcpi->getPos())//E overrules C so get next C item as well
                    cpi = cpl->getNextItem();
                pcpi = pcpl->getNextItem();//this item is OK -> get next one
            }
        } else {
            if (!doesTightLink(pcpi->getItem()->getBestChar(), pcpi->getPos(), cpl2, &cpi2))
                return 0;
            pcpi = pcpl->getNextItem();//this item is OK -> get next one
        }
    }
    return 1;
}


int Classifier::doesLink(char chr, int pos, CharPosList *cpl2, CharPosItem **cpi2, ProbCharPosList *pcpl2,
                         ProbCharPosItem **pcpi2) {
    while ((*cpi2) != 0 && (*cpi2)->getPos() < pos)//look for next relevant att. in cl.C
        (*cpi2) = cpl2->getNextItem();
    if ((*cpi2) == 0 || (*cpi2)->getPos() > pos) {//next specific att. in cl.C comes later
        while ((*pcpi2) != 0 && (*pcpi2)->getPos() < pos)//look for next relevant specific att. in cl.E
            (*pcpi2) = pcpl2->getNextItem();
        if ((*pcpi2) != 0 && (*pcpi2)->getPos() == pos) {
            if ((*pcpi2)->getItem()->getBestChar() == chr)//anticpates no change -> cannot link
                return 0;
        }
    } else {//given: pos==(*cpi2)->getPos()
        if ((*cpi2)->getChar() != chr) // does not match -> cannot link
            return 0;
    }
    return 1;
}


int Classifier::doesTightLink(char chr, int pos, CharPosList *cpl2, CharPosItem **cpi2) {
    while ((*cpi2) != 0 && (*cpi2)->getPos() < pos)//look for next relevant att. in cl.C
        (*cpi2) = cpl2->getNextItem();
    if ((*cpi2) == 0 || (*cpi2)->getPos() > pos) {//no tight link since condition does not specify required position
        return 0;
    } else {//given: pos==(*cpi2)->getPos()
        if ((*cpi2)->getChar() != chr) // does not match -> cannot link
            return 0;
    }
    return 1;
}


/**
 * Returns a new classifier that merges this one and cl2. 
 * Conditions are specialized in all attributes that are specialized in this condition or the condition of cl2.
 * Effects are merged.
 * The mark of the new classifier stays empty.
 */
Classifier *Classifier::mergeClassifiers(Classifier *cl2, Perception *percept, int time) {
    Classifier *newCl = new Classifier();

    newCl->C = new Condition(C);
    newCl->C->specialize(cl2->C);

    newCl->A = new Action(A);

    newCl->E = new Effect(E, cl2->E, q, cl2->q, percept);

    newCl->M = new PMark();

    newCl->r = (r + cl2->r) / 2.;
    newCl->q = (q + cl2->q) / 2.;
    if (newCl->q < Q_INI)
        newCl->q = Q_INI;
    newCl->num = 1;
    newCl->tga = time;
    newCl->talp = time;
    newCl->tav = 0;
    newCl->exp = 1;
    newCl->ee = 0;

    return newCl;
}


/**
 * Executes the generalizing mutation in the classifier.
 */
void Classifier::mutate() {
    int size = C->getSpecificity();
    for (int i = 0; i < size; i++) {
        if (frand() < MU) {
            assert(C->generalize(i) == 1);
            i--;
            size--;
        }
    }
}

/**
 * Executes crossover (chooses the crossover type).
 * Note that this method actually executes crossover. If crossover should be applied is checked elsewhere.
 */
void Classifier::crossover(Classifier *cl2) {
    switch (X_TYPE) {
        case 0:
            C->uniformCrossover(cl2->C);
            break;
        case 1:
            C->onePointCrossover(cl2->C, Perception::length);
            break;
        case 2:
            C->twoPointCrossover(cl2->C, Perception::length);
            break;
        default:
            cout << "X_TYPE is not set correctly!!!" << endl;
            return;
    }
    q = (q + cl2->q) / 2.;
    r = (r + cl2->r) / 2.;
    cl2->q = q;
    cl2->r = r;
}

/**
 * Creates a classifier with specified condition, action, and effect.
 */
Classifier::Classifier(Condition *con, Action *act, Effect *eff) {
    initialize(con, act, eff, new PMark(), 0, Q_INI, R_INI, IR_INI, AVT_INI);
}

/**
 * Creates a classifier with specified condition, action, effect, and time stamps.
 */
Classifier::Classifier(Condition *con, Action *act, Effect *eff, int time) {
    initialize(con, act, eff, new PMark(), time, Q_INI, R_INI, IR_INI, AVT_INI);
}

/**
 * Creates an even more initialized classifier.
 */
Classifier::Classifier(Condition *con, Action *act, Effect *eff, int time, double quality, double rewardPrediction,
                       double imRePrediction) {
    initialize(con, act, eff, new PMark(), time, quality, rewardPrediction, imRePrediction, AVT_INI);
}

/**
 * Initializes and sets the specified classifier parameters.
 */
void Classifier::initialize(Condition *con, Action *act, Effect *eff, PMark *mark, int time, double qual, double rew,
                            double imRew, double avApplTime) {
    C = con;
    A = act;
    E = eff;
    M = mark;
    q = qual;
    r = rew;
    i = imRew;
    num = 1;
    tga = time;
    talp = time;
    tav = avApplTime;
    exp = 1;
    ee = 0;
}

/**
 * Sets the ALP time stamp and the application average parameter.
 */
double Classifier::setALPTimeStamp(int time) {
    if (1. / exp > BETA)
        tav = (tav * exp + (time - talp)) /
              (exp + 1);
    else
        tav += BETA * ((time - talp) - tav);
    talp = time;
    return tav;
}

/**
 * Controls if the classifier satisfies the subsume criteria.
 */
int Classifier::isSubsumer() {
    if (exp > THETA_EXP)
        if (q > THETA_R)
            if (M->isEmpty())
                return 1;
    return 0;
}

/**
 * Returns if the classifier is marked.
 */
int Classifier::isMarked() {
    if (M->isEmpty())
        return 0;
    return 1;
}

/**
 * Returns if the classifier is formally more general than classifier 'cl'.
 */
int Classifier::isMoreGeneral(Classifier *cl) {
    if (C->getSpecificity() < cl->C->getSpecificity())
        return 1;
    return 0;
}


ostream &operator<<(ostream &out, Classifier *cl) {
    out << cl->C << " " << cl->A << " " << cl->E << " ";
    out << cl->M << " ";
    out << cl->q << "\t" << cl->r << "\t" << cl->i << "\t" << cl->exp << " " << cl->tga << "\t" << cl->talp << " "
        << cl->tav << " " << cl->num;
    return out;
}

File: alcs/acs2/Classifier.py
from random import random, sample, randint

from alcs import Perception
from alcs.acs2 import Condition, Effect, PMark, ACS2Configuration


class Classifier(object):
    def __init__(self,
                 condition=None,
                 action=None,
                 effect=None,
                 quality=0.5,
                 reward=0.5,
                 intermediate_reward=0,
                 numerosity=1,
                 experience=1,
                 talp=None,
                 tga=0,
                 tav=0,
                 cfg=None):

        if cfg is None:
            raise TypeError("Configuration should be passed to Classifier")

        self.cfg = cfg

        self.condition = Condition(condition, cfg=self.cfg) \
            if condition is not None else Condition(cfg=self.cfg)
        self.action = action if action is not None else None
        self.effect = Effect(effect, cfg=self.cfg) \
            if effect is not None else Effect(cfg=self.cfg)
        self.mark = PMark(cfg=self.cfg)

        # Quality - measures the accuracy of the anticipations
        self.q = quality

        # The reward prediction - predicts the reward expected after
        # the execution of action A given condition C
        self.r = reward

        # Intermediate reward
        self.ir = intermediate_reward

        # Numerosity
        self.num = numerosity

        # Experience
        self.exp = experience

        # When ALP learning was triggered
        self.talp = talp

        self.tga = tga

        # Application average
        self.tav = tav

        # I don't know yet what it is
        self.ee = 0

    def q3num(self):
        return pow(self.q, 3) * self.num

    def __eq__(self, other):
        return self.is_similar(other)

    def __hash__(self):
        return hash((str(self.condition), self.action, str(self.effect)))

    def __repr__(self):
        return "{}-{}-{} @ {}".format(self.condition,
                                      self.action,
                                      self.effect,
                                      hex(id(self)))

    @classmethod
    def copy_from(cls, old_cls, time):
        """
        Copies old classifier with given time (tga, talp).
        Old tav gets replaced with new value.
        New classifier also has no mark.

        :param old_cls: classifier to copy from
        :param time: time of creation
        :return: new classifier
        """
        new_cls = cls(
            condition=Condition(old_cls.condition, old_cls.cfg),
            action=old_cls.action,
            effect=old_cls.effect,
            quality=old_cls.q,
            reward=old_cls.r,
            intermediate_reward=old_cls.ir,
            cfg=old_cls.cfg)

        new_cls.tga = time
        new_cls.talp = time
        new_cls.tav = old_cls.tav

        return new_cls

    @classmethod
    def cover_triple(cls,
                     previous_situation: Perception,
                     action: int,
                     situation: Perception,
                     time: int,
                     cfg: ACS2Configuration):
        """
        Creates a classifier that anticipates a change correctly

        :param previous_situation:
        :param action:
        :param situation:
        :param time:
        :param cfg: configuration

        :return: new classifier
        """
        new_cl = cls(action=action, cfg=cfg)  # TODO: p5 exp=0, r=0 (paper)
        new_cl.tga = time
        new_cl.talp = time

        new_cl.specialize(previous_situation, situation)

        return new_cl

    @property
    def fitness(self):
        return self.q * self.r

    @property
    def specified_unchanging_attributes(self):
        """
        Determines the number of specified unchanging attributes in
        the classifier. An unchanging attribute is one that is anticipated
        not to change in the effect part.

        :return: number of specified unchanging attributes
        """
        spec = 0

        for cpi, epi in zip(self.condition, self.effect):
            if cpi != self.cfg.classifier_wildcard and \
                epi == self.cfg.classifier_wildcard:
                spec += 1

        return spec

    @property
    def specificity(self):
        return self.condition.specificity / len(self.condition)

    def does_anticipate_change(self) -> bool:
        """
        :return: true if the effect part contains any specified attributes
        """
        return self.effect.number_of_specified_elements > 0

    def is_reliable(self):
        return self.q > self.cfg.theta_r

    def is_inadequate(self):
        return self.q < self.cfg.theta_i

    def update_reward(self, p: float) -> float:
        self.r += self.cfg.beta * (p - self.r)
        return self.r

    def update_intermediate_reward(self, rho) -> float:
        self.ir += self.cfg.beta * (rho - self.ir)
        return self.ir

    def increase_experience(self) -> int:
        self.exp += 1
        return self.exp

    def increase_quality(self) -> float:
        self.q += self.cfg.beta * (1 - self.q)
        return self.q

    def decrease_quality(self) -> float:
        self.q -= self.cfg.beta * self.q
        return self.q

    def specialize(self,
                   previous_situation: Perception,
                   situation: Perception):
        """
        Specializes the effect part where necessary to correctly anticipate
        the changes from p0 to p1 and returns a condition which specifies
        the attributes which must be specified in the condition part.
        The specific attributes in the returned conditions are set to
        the necessary values.

        :param previous_situation:
        :param situation:
        """
        for idx, item in enumerate(situation):
            if previous_situation[idx] != situation[idx]:
                self.effect[idx] = situation[idx]
                self.condition[idx] = previous_situation[idx]

    def predicts_successfully(self,
                              p0: Perception,
                              action: int,
                              p1: Perception) -> bool:
        """
        Check if classifier matches previous situation `p0`,
        has action `action` and predicts the effect `p1`

        :param p0: previous situation
        :param action:
        :param p1: anticipated situation after execution action
        :return: True if classifier makes successful predictions,
        False otherwise
        """
        if self.condition.does_match(p0):
            if self.action == action:
                if self.does_anticipate_correctly(p0, p1):
                    return True

        return False

    def does_anticipate_correctly(self,
                                  previous_situation: Perception,
                                  situation: Perception) -> bool:
        """
        Returns if the classifier correctly anticipates the change from previous_situation to situation.
        :param previous_situation:
        :param situation:
        :return:
        """
        return self.effect.does_anticipate_correctly(
            previous_situation, situation)

    def set_mark(self, perception: Perception) -> None:
        """
        Marks classifier with given perception taking into consideration its
        condition.

        Specializes the mark in all attributes which are not specified
        in the conditions, yet

        :param perception: current situation
        """
        if self.mark.set_mark_using_condition(self.condition, perception):
            self.ee = 0

    def set_alp_timestamp(self, time: int) -> None:
        """
        Sets the ALP time stamp and the application average parameter.

        :param time: current step
        """
        # TODO p5: write test
        if 1. / self.exp > self.cfg.beta:
            self.tav = (self.tav * self.exp + (time - self.talp)) / (
                self.exp + 1)
        else:
            self.tav += self.cfg.beta * ((time - self.talp) - self.tav)

        self.talp = time

    def expected_case(self, perception: Perception, time: int):
        """
        Controls the expected case of a classifier. If the classifier
        is to specific it tries to add some randomness to it by
        generalizing some attributes.

        :param perception:
        :param time:
        :return: new classifier or None
        """
        diff = self.mark.get_differences(perception)

        if diff is None:
            self.increase_quality()
            return None

        no_spec = self.specified_unchanging_attributes
        no_spec_new = diff.specificity
        child = self.copy_from(self, time)

        if no_spec >= self.cfg.u_max:
            while no_spec >= self.cfg.u_max:
                res = self.generalize_unchanging_condition_attribute(no_spec)
                assert res is True
                no_spec -= 1

            while no_spec + no_spec_new > self.cfg.u_max:
                if random() < 0.5:
                    diff_idx = randint(0, no_spec_new)
                    diff.generalize(diff_idx)
                    no_spec_new -= 1
                else:
                    if self.generalize_unchanging_condition_attribute(no_spec):
                        no_spec -= 1
        else:
            while no_spec + no_spec_new > self.cfg.u_max:
                diff_idx = randint(0, no_spec_new)
                diff.generalize(diff_idx)
                no_spec_new -= 1

        child.condition.specialize(new_condition=diff)

        if child.q < 0.5:
            child.q = 0.5

        return child

    def unexpected_case(self,
                        previous_perception: Perception,
                        perception: Perception,
                        time: int):
        """
        Controls the unexpected case of the classifier.

        :param previous_perception:
        :param perception:
        :param time:
        :return: specialized classifier if generation was possible,
        None otherwise
        """
        self.decrease_quality()
        self.set_mark(previous_perception)

        # Return if the effect is not specializable
        if not self.effect.is_specializable(previous_perception, perception):
            return None

        child = self.copy_from(self, time)

        # TODO: p5 maybe also take into consideration cl.E = # (paper)
        child.specialize(previous_perception, perception)

        if child.q < 0.5:
            child.q = 0.5

        return child

    def mutate(self, randomfunc=random):
        """
        Executes the generalizing mutation in the classifier.
        """
        for idx, cond in enumerate(self.condition):
            if cond != self.cfg.classifier_wildcard and \
                randomfunc() < self.cfg.mu:
                self.condition.generalize(idx)

    def is_similar(self, other) -> bool:
        """
        Check if classifier is equals to `other` classifier in condition,
        action and effect part.

        :param other: other classifier
        :return: True if equals, False otherwise
        """
        if self.condition == other.condition and \
            self.action == other.action and \
            self.effect == other.effect:
            return True
        return False

    def is_more_general(self, other) -> bool:
        """
        Checks if the classifier is formally more general than `other`.

        :param other: other classifier to compare
        :return: True if `other` classifier is more general
        """
        if self.condition.specificity < other.condition.specificity:
            return True

        return False

    def generalize_unchanging_condition_attribute(
        self, no_spec, randomfunc=randint) -> bool:
        """
        Generalizes one randomly unchanging attribute in the condition.
        An unchanging attribute is one that is anticipated not to change
        in the effect part.
        :param no_spec: number of unchanging attributes
        :param randomfunc: specifies random function for distinguishing
        which attribute to generalize
        :return: True if attribute was generalized, False otherwise
        """
        if no_spec == 0:
            return False  # nothing to generalize

        att_idx = randomfunc(0, no_spec - 1)  # id of attribute to generalize
        pos = 0  # current unchanging attribute id

        for idx, (cpi, epi) in enumerate(zip(self.condition, self.effect)):
            if cpi != self.cfg.classifier_wildcard and \
                epi == self.cfg.classifier_wildcard:

                if att_idx == pos:
                    self.condition.generalize(idx)
                    return True
                else:
                    pos += 1

        return False

    def does_subsume(self, other) -> bool:
        """
        Returns if a classifier subsumes `other` classifier

        :param other: other classifiers
        :return: True if `other` classifier is subsumed, False otherwise
        """
        if self._is_subsumer() and \
            self.is_more_general(other) and \
            self.condition.does_match(other.condition) and \
            self.effect == other.effect:
            return True

        return False

    def _is_subsumer(self) -> bool:
        """
        Controls if the classifier satisfies the subsumer criteria.

        :return: True is classifier can be considered as subsumer,
        False otherwise
        """
        if self.exp > self.cfg.theta_exp:
            if self.is_reliable():
                if self.mark.is_empty():
                    return True

        return False

    def is_unmarked(self):
        return not self.is_marked()

    def is_marked(self):
        """
        Returns if classifier is marked
        """
        if self.mark.is_empty():
            return False

        return True

    def crossover(self, cl2, samplefunc=sample):
        """
        Executes crossover with other classifier
        :param samplefunc:
        :param cl2: other classifier
        """
        self.condition.two_point_crossover(
            cl2.condition, samplefunc=samplefunc)

        q = float(sum([self.q, cl2.q]) / 2)
        r = float(sum([self.r, cl2.r]) / 2)

        cl2.q = q
        cl2.r = r

    def does_match(self, situation):
        """
        Returns if the classifier matches the situation.
        :param situation:
        :return:
        """
        return self.condition.does_match(situation)

    def does_match_backwards(self, situation):
        """
        Returns if 'situation' is matched by the anticipations.
        This is only the case if the specified conditions that have #-symbols in the effect part are also matched!
        :param situation:
        :return:
        """
        p = self.condition.get_backwards_anticipation(situation)
        if self.effect.does_match(situation, p):
            return True
        return False

    def get_best_anticipation(self, perception):
        """
        Returns the anticipation, the classifier believes to happen most probably.
        This is usually the normal anticipation. However, if PEEs are activated, the most probable
        value of each attribute is returned.
        :param perception: Perception
        :return:
        """
        return self.effect.get_best_anticipation(perception)

    def get_backwards_anticipation(self, perception):
        """
        Returns the backwards anticipation.
        Returns -1 if the backwards anticipation was impossible to create. This is the case if
        changing attributes are not specified in the conditions.
        :param perception:
        :return:
        """
        back_anticipation = self.condition.get_backwards_anticipation(perception)
        if not self.effect.does_specify_only_changes_backwards(back_anticipation, perception):
            # If a specified attribute in the effect part matches the anticipated 'back_anticipation',
            # the backward anticipation fails, because a specified attribute in Effect means a change!
            return None
        return back_anticipation
